src/libraries/System.Reflection.Metadata/src/System/Reflection/Metadata/TypeNameParserHelpers.cs

```diff
@@ -102,29 +102,41 @@ static int GetUnescapedOffset(ReadOnlySpan<char> input, int startOffset)
             static bool NeedsEscaping(char c) => c is '[' or ']' or '&' or '*' or ',' or '+' or EscapeCharacter;
         }
 
-        internal static ReadOnlySpan<char> GetName(ReadOnlySpan<char> fullName)
+        internal static ReadOnlySpan<char> GetNamespace(ReadOnlySpan<char> fullName)
         {
-            // The two-value form of MemoryExtensions.LastIndexOfAny does not suffer
-            // from the behavior mentioned in the comment at the top of GetFullTypeNameLength.
-            // It always takes O(m * i) worst-case time and is safe to use here.
+            int offset = fullName.LastIndexOf('.');
+            return offset < 0 ? [] : fullName.Slice(0, offset);
+        }
+
+        internal static ReadOnlySpan<char> GetName(ReadOnlySpan<char> fullName, bool isNested)
+        {
+            // If the type is nested, return the whole name after the plus sign.
+            int offset = isNested
+                ? LastIndexOfUnescaped(fullName, '+')
+                : fullName.LastIndexOf('.');
+
+            return offset < 0 ? fullName : fullName.Slice(offset + 1);
+        }
 
-            int offset = fullName.LastIndexOfAny('.', '+');
+        private static int LastIndexOfUnescaped(ReadOnlySpan<char> str, char c)
+        {
+            int offset = str.LastIndexOf(c);
 
-            if (offset > 0 && fullName[offset - 1] == EscapeCharacter) // this should be very rare (IL Emit & pure IL)
+            if (offset > 0 && str[offset - 1] == EscapeCharacter) // this should be very rare (IL Emit & pure IL)
             {
-                offset = GetUnescapedOffset(fullName, startIndex: offset);
+                offset = LastIndexOfAnyUnescapedSlow(str, c, startIndex: offset);
             }
 
-            return offset < 0 ? fullName : fullName.Slice(offset + 1);
+            return offset;
 
-            static int GetUnescapedOffset(ReadOnlySpan<char> fullName, int startIndex)
+            static int LastIndexOfAnyUnescapedSlow(ReadOnlySpan<char> str, char c, int startIndex)
             {
                 int offset = startIndex;
                 for (; offset >= 0; offset--)
                 {
-                    if (fullName[offset] is '.' or '+')
+                    if (str[offset] == c)
                     {
-                        if (offset == 0 || fullName[offset - 1] != EscapeCharacter)
+                        if (offset == 0 || str[offset - 1] != EscapeCharacter)
```

Comment:
This does not look correct for a type name like: `NameEndingInBackSlash\\+NestedName`

Instead of looking for `+` and figuring out whether it is escaped, would it be better to just get the length of the declaring type typename and slice it from there?

Reply:
Good idea, done and added your test case.